test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp

```diff
@@ -346,23 +347,64 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
   // Communicators might be aborted here, further operations would fail.
 }
 
+// Function to read what we wrote to the local disk for validation.
+std::string readTraceInfoFromLocalDisk(int rank) {
+  auto filename = c10::str("/tmp/test_nccl_trace_rank_", rank);
+  std::ifstream file(filename, std::ios::binary);
+  size_t size;
+  // Read the strings from the file
+  if (file) { // While the file stream is in good state
+    // Read the size of the string
+    file.read(reinterpret_cast<char*>(&size), sizeof(size));
+    if (file) { // Check if read was successful to proceed
+      std::string str(size, '\0'); // Initialize string with 'size'
+      file.read(&str[0], size);
+      if (file) {
+        return str;
+      }
+    }
+  }
+  return "";
+}
+
 TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNoHeartbeat) {
   if (skipTest()) {
     return;
   }
 
   int heartBeatIntervalInSec = 2;
+  std::vector<char> traces;
   std::string timeInterval = std::to_string(heartBeatIntervalInSec);
   ASSERT_TRUE(setenv(c10d::NCCL_BLOCKING_WAIT, "1", 1) == 0);
   ASSERT_TRUE(
       setenv(c10d::TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC, timeInterval.c_str(), 1) ==
       0);
   ASSERT_TRUE(setenv(c10d::TORCH_NCCL_ENABLE_MONITORING, "1", 1) == 0);
+  // Enable nccl flight recorder.
+  ASSERT_TRUE(setenv("TORCH_NCCL_TRACE_BUFFER_SIZE", "10", 1) == 0);
   auto options = c10d::ProcessGroupNCCL::Options::create();
   // Set a long watchdog timeout, so that we have enough time to lock the
   // watchdog and let the heartbeat monitor thread to kick in.
   options->timeout = std::chrono::milliseconds(30000);
   ProcessGroupNCCLNoHeartbeatCaught pg(store_, 0, 1, options);
+  // The lamda function here is very similar to the default callback.
+  // The only difference is that we are storing the nccl trace in a vec<char>
+  // so that we can use it for validation.
+  pg.registerDebugInfoCallbackStorer(
+      [&](int rank, const std::string& ncclTrace) {
+        auto filename = c10::str("/tmp/test_nccl_trace_rank_", rank);
+        // We override the file everytime we write to it.
+        std::ofstream file(
+            filename, std::ios::out | std::ios::binary | std::ios::trunc);
+        // Write the size of the string followed by the string itself.
+        // This can help in retrieving the strings back if needed.
+        size_t size = ncclTrace.size();
+        // std::strncpy or std::strcpy does not work here because there is
+        // null terminator character (\0) in the string.
+        traces.assign(ncclTrace.begin(), ncclTrace.end());
+        file.write(reinterpret_cast<const char*>(&size), sizeof(size));
+        file.write(ncclTrace.data(), size);
```

Comment:
shall we call `file.flush()` after writing?

Reply:
actually, since right after this, the destructor of `ofstream` will be called. The `flush` will be called automatically.